### PR TITLE
Follow Up for Add source fallback for keyword fields

### PR DESCRIPTION
**Pull request by jdconrad**
Original PR here (https://github.com/elastic/elasticsearch/pull/87765). This PR adds updates to add source fallback through a `MappedType#scriptDatafieldBuilder` method along with all the required additional plumbing for the new scripting fields API to use. This change is based on the design discussion results outlined here (https://github.com/elastic/elasticsearch/issues/80504#issuecomment-1165959827)

Note that the `scriptDatafieldBuilder` method returns a `Tuple<Boolean, IndexFieldData.Builder>` where the `Boolean` value is based on whether or not we used a fallback to retrieve values. This is required for the old-style script access through `doc['field']` to maintain its current user-facing values. The `LeafDocLookup` cache is a reflection of this where additional plumbing has been added so that the new-style and old-style api can share the cached values retrieved from doc values, but only the new-style api will retrieve values from source.

I would please request that a closer look be given to the changes in `IndexFieldService` related to caching as I wasn't sure if we should be caching any values retrieved from source fallback via runtime fields. I also would please request that attention be given to cyclic runtime field checking as while I think the new path covers it correctly, I wasn't sure if I missed some corner cases.

### CODE REVIEW

**Comment by jdconrad** [comment]
I have made the following updates based on discussions with the search team members:

1. The keyword field type now uses a `ValueFetcher` to do source fallback instead of a runtime field. The majority of this new code is contained in `KeywordValueFetcherIndexFieldData`. This is an example of how we could apply value fetchers to other fields required for the new scripting API as well. There is probably opportunity to extract out code, but I didn't want to over-complicate anything until it was looked at.
2. I removed the now extraneous code from runtime fields. This gets rid of source paths having to be plumbed into them.
3. I removed the `Tuple` return from `scriptFielddataBuilder` and instead replaced this information with a marker interface on the `KeywordValueFetcherDocValues`. Since we are no longer using runtime fields it's much simpler to just add a marker interface to any of the new doc values generated from value fetcher to use an instance of check against in `LeafDocLookup` to differentiate between the old doc style and the new scripting fields style.

@javanna Would you please take a look and let me know what you think when you have a bit of time? Thanks!

**Review by javanna — COMMENTED** [review]
I left some high-level comments, and asked @romseygeek to have a better look and get in touch. Thanks for iterating on this!

File: server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class KeywordValueFetcherIndexFieldData
+    implements
+        IndexFieldData<KeywordValueFetcherIndexFieldData.KeywordValueFetcherLeafFieldData> {
+
+    public static class Builder implements IndexFieldData.Builder {
+        private final String name;
+        private final ValueFetcher valueFetcher;
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java:41]: One doubt I have around using value fetchers: we have been talking about allowing value fetchers to load from doc_values or stored fields. Should we make the fielddata that's based on value fetcher depend directly on the source value fetcher then? Would it make any sense to ever load from doc_values through this fielddata impl? Would like to gather feedback about this.

File: server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java [hunk 2]
```diff
@@ -693,21 +692,26 @@ public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName, S
...
-            return new Tuple<>(
-                false,
-                new StringScriptFieldData.Builder(
-                    name(),
-                    StringFieldScript.PARSE_FROM_SOURCE_PATHS.newFactory(name(), Collections.emptyMap(), searchLookup.get()),
-                    KeywordDocValuesField::new
-                )
+            return new KeywordValueFetcherIndexFieldData.Builder(
+                name(),
+                new SourceValueFetcher(searchLookup.get().sourcePaths(name()), nullValue) {
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java:702]: The ones currently created for source depend on receiving an existing `SearchExecutionContext`. I would definitely be open an an alternative for this, but it didn't seem correct to try to pass the `SearchExecutionContext` through to scripting and then through to a mapped type.

File: server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+        @Override
+        public int docValueCount() {
+            return values.size();
+        }
+
+        @Override
+        public BytesRef nextValue() throws IOException {
+            assert iterator.hasNext();
+            return new BytesRef(iterator.next().toString());
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java:203]: I'm definitely open to improvement here. I just wasn't sure of the best way to get the required BytesRef for this specific type of doc values. I like the idea you mentioned here of possibly re-using value fetchers for gathering source from runtime fields, but I imagine that could be a small project on its own. I'll take a look for the cast within the mapped field type to see if that could work.

File: server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+import org.elasticsearch.search.sort.BucketedSort;
+import org.elasticsearch.search.sort.SortOrder;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.SortedSet;
+import java.util.TreeSet;
+
+public class KeywordValueFetcherIndexFieldData
```
[the remark this answers is not in the document]
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java:35]: There is likely some common ground here. I agree we should look for a way to share this if possible. 

We could probably re-name and re-use `BinaryScriptFieldData` as the common base class for both `KeywordValueFetcherIndexFieldData` and `StringScriptFieldData`.

I also like the idea of routing runtime fields using source through this path instead. One caveat is the value fetchers work on source paths and parent fields as opposed to a single user-provided path so we would have to workout if that's an okay change for the user.

File: server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java (hunk 1 above)
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/index/fielddata/KeywordValueFetcherIndexFieldData.java:41]: That's a good question. This seems like a good topic for a discussion w/ the search team.

**Comment by jdconrad** [comment]
I spoke with @romseygeek, and I'm going to add a couple more examples of how this would work for a numeric type (like long) and a structured type (like GeoPoint) to see if this is a good path to continue forward with.

**Comment by jdconrad** [comment]
@romseygeek I added source fallback for integer and geo point type fields. I also abstracted out the source value fetcher index field data code to try to remove some of the boiler plate. I only added what was strictly necessary for scripting for now, but I believe the code would be easily extendable to other features that wanted to take advantage by making different types of emulated doc values available. I did not do any of the other numeric type fields, but they would follow basically the same pattern. 

Would you please take another pass when you have a bit of time? Thanks!

Also after considering this code for a while I realize it was really only designed to have doc values from Lucene as a direct source. With the addition of runtime fields and now possibly source fallback, I do wonder if there's a way we could abstract out the source of the doc values and pass it into the index field data so index field data can be shared when the source is actual doc values and when the source is emulated doc values instead of needing a separate one for each. But this is a discussion for a different time.

**Comment by jdconrad** [comment]
After speaking with @romseygeek I'm going to produce another WIP PR where instead of having two fielddataBuilder methods (one for scripting and one for search), we're going to have a single method that has options on it to see if that reduces plumbing overhead.

**Comment by jdconrad** [comment]
Closing this in favor of https://github.com/elastic/elasticsearch/pull/88735

**Pull request closed without merging**